lua/sim/Matchstate/AbstractVictoryCondition.lua

```diff
@@ -0,0 +1,230 @@
+--******************************************************************************************************
+
+local DebugComponent = import("/lua/shared/components/DebugComponent.lua").DebugComponent
+local SyncGameResult = import("/lua/simsyncutils.lua").SyncGameResult
+
+-- upvalue for performance
+local TableGetn = table.getn
+local TableInsert = table.insert
+
+---@class AbstractVictoryCondition : DebugComponent, Destroyable
+---@field Trash TrashBag
+AbstractVictoryCondition = Class(DebugComponent) {
+
+    ---@param self AbstractVictoryCondition
+    __init = function(self)
+        self.Trash = TrashBag()
+    end,
+
+    ---@param self AbstractVictoryCondition
+    Destroy = function(self)
+        self:OnDestroy()
+    end,
+
+    ---@param self AbstractVictoryCondition
+    OnDestroy = function(self)
+        self.Trash:Destroy()
+    end,
+
+    --- A utility function that returns whether a unit is eligible to be considered in the game for the victory condition.
+    ---@param self AbstractVictoryCondition
+    ---@param unit Unit
+    ---@return boolean
+    UnitIsEligible = function(self, unit)
+
+        -- this can happen occasionally, note that we explicitly do **not** use the `.Dead` property! The
+        -- use of that property depends on the execution order in Lua. It is deterministic, but it will
+        -- generate strange results when for example two ACUs die in the same tick. Using `.Dead` will
+        -- re-introduce the 'draw bug' that bugged the community for years.
+
+        if IsDestroyed(unit) then
+            return false
+        end
+
+        -- only units that are finished are taken into account.
+
+        if unit:GetFractionComplete() <= 1.0 then
+            return false
+        end
+
+        -- only units that are not recalling are taken into account.
+
+        if unit.RecallingAfterDefeat then
+            return false
+        end
+
+        return true
+    end,
+
+    --- A utility function that returns whether the given brain has any eligible units of the given category.
+    ---@param self AbstractVictoryCondition
+    ---@param aiBrain AIBrain
+    ---@param categories EntityCategory
+    ---@return boolean
+    BrainHasEligibleUnits = function(self, aiBrain, categories)
+        local units = aiBrain:GetListOfUnits(categories, false)
+        if (units) then
+            for k = 1, TableGetn(units) do
+                local unit = units[k]
+                if self:UnitIsEligible(unit) then
+                    return true
+                end
+            end
+        end
+
+        return false
+    end,
+
+    --- A utility function that retrieves all brains that are still participating in the match.
+    ---@param self AbstractVictoryCondition
+    ---@return AIBrain[]
+    GetArmyBrains = function(self)
+        local participatingArmyBrains = {}
+        for k, aiBrain in ArmyBrains do
+            if not (ArmyIsOutOfGame(aiBrain:GetArmyIndex() or aiBrain:IsDefeated())) then
```

Comment:
- `ArmyIsOutOfGame` had its parentheses around the entire `or` statement
- `IsDefeated` already checks `ArmyIsOutOfGame`:
https://github.com/FAForever/fa/blob/e16b829ee2603450988f283e7ee1776026c53061/lua/aibrain.lua#L411-L417
```suggestion
            if not aiBrain:IsDefeated() then
```